test/functional/test-gesture-recognizers.js

```diff
@@ -596,3 +596,255 @@ describe('TapzoomRecognizer', () => {
+              data.centerClientX == 100 &&
+              data.centerClientY == 100 &&
+              data.dir == 1 &&
+              data.deltaX == 10 &&
+              data.deltaY == 10 &&
+              data.time == 11 &&
+              diff(data.velocityX, 0.79, 1e-2) &&
+              diff(data.velocityY, 0.79, 1e-2));
+        }), null).once();
+    gesturesMock.expects('signalEnd_').never();
+
+    clock.tick(10);
+    recognizer.acceptStart();
+
+    expect(recognizer.eventing_).to.equal(true);
+    expect(recognizer.startTime_).to.equal(1);
+    expect(recognizer.startX1_).to.equal(90);
+    expect(recognizer.startY1_).to.equal(80);
+    expect(recognizer.startX2_).to.equal(110);
+    expect(recognizer.startY2_).to.equal(120);
+    expect(recognizer.lastX1_).to.equal(80);
+    expect(recognizer.lastY1_).to.equal(70);
+    expect(recognizer.lastX2_).to.equal(120);
+    expect(recognizer.lastY2_).to.equal(130);
+    expect(recognizer.prevTime_).to.equal(11);
+    expect(recognizer.prevDeltaX_).to.equal(20);
+    expect(recognizer.prevDeltaY_).to.equal(20);
+  });
+
+  it('should emit on touchmove after start', () => {
+    clock.tick(1);
+    recognizer.onTouchStart({touches:
+        [{clientX: 90, clientY: 80},
+         {clientX: 110, clientY: 120}]});
+    recognizer.onTouchMove({touches:
+        [{clientX: 80, clientY: 70},
+         {clientX: 120, clientY: 130}]});
+    gesturesMock.expects('signalEmit_').withExactArgs(recognizer,
+        sinon.match((data) => true), null).once();
+    clock.tick(10);
+    recognizer.acceptStart();
+
+    let event = {touches:
+        [{clientX: 70, clientY: 60},
+         {clientX: 130, clientY: 140}]};
+    gesturesMock.expects('signalEmit_').withExactArgs(recognizer,
+        sinon.match((data) => {
+          return (data.first === false && data.last === false &&
+              data.centerClientX == 100 &&
+              data.centerClientY == 100 &&
+              data.dir == 1 &&
+              data.deltaX == 20 &&
+              data.deltaY == 20 &&
+              data.velocityX > 0 &&
+              data.velocityY > 0);
+        }), event).once();
+    gesturesMock.expects('signalEnd_').never();
+
+    clock.tick(10);
+    let res = recognizer.onTouchMove(event);
+    expect(res).to.equal(true);
+  });
+
+  it('should stop on touchend; velocity doesn\'t change', () => {
+    clock.tick(1);
+    recognizer.onTouchStart({touches:
+        [{clientX: 90, clientY: 80},
+         {clientX: 110, clientY: 120}]});
+    recognizer.onTouchMove({touches:
+        [{clientX: 80, clientY: 70},
+         {clientX: 120, clientY: 130}]});
+    gesturesMock.expects('signalEmit_').withExactArgs(recognizer,
+        sinon.match((data) => true), null).once();
+    clock.tick(10);
+    recognizer.acceptStart();
+
+    let event = {};
+    gesturesMock.expects('signalEmit_').withExactArgs(recognizer,
+        sinon.match((data) => {
+          return (data.first === false && data.last === true &&
```

Comment:
null == undefined   // true
    false == '0'        // true
    false == 'false'    // false

but

    false == undefined  // false
    false == null       // false

So its good to use `===` to avoid acidentially type converting